### PR TITLE
#2335 Add support for session properties in model defaults

### PR DESCRIPTION
**Pull request by ciprian-cgr**
This should allow you to set session_properties on the project configuration which will then be propagated on the models. If custom properties are also set on the specific model, those will not be overwritten by the default ones. It might need some more work when adding the default props to handle quoted/unquoted identifiers better; right now, it sets them by default as quoted=False.

### CODE REVIEW

File: sqlmesh/core/model/definition.py [hunk 1]
```diff
@@ -1876,6 +1877,21 @@ def _split_sql_model_statements(
     return query, expressions[:pos], expressions[pos + 1 :]
 
 
+def _resolve_custom_session_params(
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1880]: If the purpose of the function is to only resolve session properties it feels somewhat counterintuitive to require 2 dicts in arguments and mutation one of them. How about the following signature:
```
def _resolve_custom_session_params(
    default: t.Optional[t.Dict[str, t.Any]], provided: t.Optional[exp.Expression]
) -> t.Optional[exp.Expression]:
    if default is not None and provided is not None:
        ...

```

File: sqlmesh/core/model/definition.py [hunk 2]
```diff
@@ -1876,6 +1877,21 @@ def _split_sql_model_statements(
...
+    defaults: t.Dict[str, t.Any] | None, **kwargs: t.Dict[str, t.Any]
+) -> None:
+    if kwargs.get("session_properties") and defaults and defaults.get("session_properties"):
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1883]: So what happens if there session properties in defaults but not in model definitions? Don't we still need to convert the dict into a sqlglot expression?

File: sqlmesh/core/model/definition.py [hunk 3]
```diff
@@ -1876,6 +1877,21 @@ def _split_sql_model_statements(
...
+        session_properties = kwargs["session_properties"]
+        session_props = {expr.this.name for expr in session_properties}
+        for k, v in defaults["session_properties"].items():
+            if k not in session_props:
+                new_eq = exp.EQ(
+                    this=exp.Literal(this=k), expression=exp.Column(this=exp.Identifier(this=v))
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1889]: The value can also be a sqlglot expression, that's why I'd suggest doing:
```
this=exp.Literal(this=k), expression=exp.convert(v)
```
Also CC: @georgesittas @tobymao to correct me if I'm wrong.

**Comment by izeigerman** [comment]
Make sure to run `make style` to run linters and code formatter

File: tests/core/test_model.py [hunk 4]
```diff
@@ -2698,6 +2698,47 @@ def test_model_table_properties_conflicts() -> None:
         sql_model.physical_properties
 
 
+def test_session_properties_on_model_and_project(sushi_context):
```
Comment (izeigerman) [inline tests/core/test_model.py:2701]: Let's also add a test when there are only defaults but nothing is set in the model definition.

File: sqlmesh/core/model/definition.py [hunk 5]
```diff
@@ -1876,6 +1880,28 @@ def _split_sql_model_statements(
...
+    provided: t.Optional[exp.Expression] | t.Optional[t.Dict[str, t.Any]],
+) -> t.Optional[exp.Expression]:
+    if provided is not None and isinstance(provided, dict):
+        session_properties = {k: exp.Literal.string(k).eq(v) for k, v in provided.items()}
+    else:
+        session_properties = (
+            {expr.this.name: expr for expr in provided} if provided is not None else {}
+        )
+
+    if defaults is not None and defaults.get("session_properties") is not None:
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1894]: Nit:
```
if defaults and "session_properties" in defaults:
```

File: sqlmesh/core/model/definition.py [hunk 6]
```diff
@@ -1876,6 +1880,28 @@ def _split_sql_model_statements(
     return query, expressions[:pos], expressions[pos + 1 :]
 
 
+def _resolve_custom_session_properties(
+    defaults: t.Optional[t.Dict[str, t.Any]],
+    provided: t.Optional[exp.Expression] | t.Optional[t.Dict[str, t.Any]],
+) -> t.Optional[exp.Expression]:
+    if provided is not None and isinstance(provided, dict):
```
Comment (izeigerman) [inline sqlmesh/core/model/definition.py:1887]: I'm pretty sure you just need `isinstance(provided, dict)` since `isinstance(None, dict)` returns false. Let me know if I'm wrong.

File: tests/core/test_model.py [hunk 7]
```diff
@@ -2698,6 +2698,76 @@ def test_model_table_properties_conflicts() -> None:
...
+        "spark.executor.memory": "1G",
+        "some_bool": True,
+        "some_float": 0.1,
+        "quoted_identifier": exp.column("quoted identifier", quoted=True),
+        "unquoted_identifier": exp.column("unquoted_identifier", quoted=False),
+        "project_level_property": "project_property",
+    }
+
+
+def test_project_level_session_properties(sushi_context):
```
Comment (izeigerman) [inline tests/core/test_model.py:2742]: Thanks for adding this test! Since you caught an edge case which involved a Python model would it make sense to add a test which tests specifically the `dict` route?

File: sqlmesh/core/model/definition.py (hunk 6 above)
Comment (ciprian-cgr) [reply sqlmesh/core/model/definition.py:1887]: Indeed, I thought it might check for an empty dict as well, though that won't cause any issue as it resolved to an empty dict anyway.

File: sqlmesh/core/model/definition.py (hunk 5 above)
Comment (ciprian-cgr) [reply sqlmesh/core/model/definition.py:1894]: yes, that was my preference too, I just thought you might have some convention I'm not aware of based on this comment 
image

**Review by izeigerman — APPROVED** [review]
Thanks a lot for addressing comments!
